File: 1718_construct_the_lexicographically_largest_valid_sequence/solution.go
// LeetCode 1718 - Construct the Lexicographically Largest Valid Sequence
// https://leetcode.com/problems/construct-the-lexicographically-largest-valid-sequence/

func constructDistancedSequence(n int) []int {
    ans := make([]int, 2*n-1)
    used := make([]bool, n+1)

    var backtrack func(i int) bool
    backtrack = func(i int) bool {
        for i < len(ans) && ans[i] != 0 {
            i++
        }
        if i == len(ans) {
            return true
        }
        for value := n; value >= 1; value-- {
            if used[value] {
                continue
            }
            if value == 1 {
                ans[i] = 1
                used[1] = true
                if backtrack(i + 1) {
                    return true
                }
                used[1] = false
                ans[i] = 0
            } else {
                j := i + value
                if j < len(ans) && ans[j] == 0 {
                    ans[i], ans[j] = value, value
                    used[value] = true
                    if backtrack(i + 1) {
                        return true
                    }
                    used[value] = false
                    ans[i], ans[j] = 0, 0
                }
            }
        }
        return false
    }

    backtrack(0)
    return ans
}

File: 1718_construct_the_lexicographically_largest_valid_sequence/solution.c
// LeetCode 1718 - Construct the Lexicographically Largest Valid Sequence
// https://leetcode.com/problems/construct-the-lexicographically-largest-valid-sequence/

#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

static bool backtrack(int i, int n, int size, int* ans, bool* used) {
    while (i < size && ans[i] != 0) {
        i++;
    }
    if (i == size) {
        return true;
    }
    for (int value = n; value >= 1; value--) {
        if (used[value]) {
            continue;
        }
        if (value == 1) {
            ans[i] = 1;
            used[1] = true;
            if (backtrack(i + 1, n, size, ans, used)) {
                return true;
            }
            used[1] = false;
            ans[i] = 0;
        } else {
            int j = i + value;
            if (j < size && ans[j] == 0) {
                ans[i] = value;
                ans[j] = value;
                used[value] = true;
                if (backtrack(i + 1, n, size, ans, used)) {
                    return true;
                }
                used[value] = false;
                ans[i] = 0;
                ans[j] = 0;
            }
        }
    }
    return false;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* constructDistancedSequence(int n, int* returnSize) {
    int size = 2 * n - 1;
    int* ans = (int*)calloc(size, sizeof(int));
    bool* used = (bool*)calloc(n + 1, sizeof(bool));
    backtrack(0, n, size, ans, used);
    free(used);
    *returnSize = size;
    return ans;
}
